File: slice-results-centerlines/c++/Graphics.cpp

#include "Graphics.h" 

#include <vtkCellData.h>
#include <vtkCellPicker.h>
//
#include <vtkExtractSelectedPolyDataIds.h>
#include <vtkExtractSelection.h>
//
#include <vtkGenericCell.h>
//
#include <vtkNamedColors.h>
#include <vtkPointData.h>
//
#include <vtkSelection.h>
#include <vtkSelectionNode.h>
//
#include <vtkUnstructuredGrid.h>

//----------
// Graphics
//----------
//
Graphics::Graphics()
{
  m_Mesh = nullptr;

  // Add a renderer.
  m_Renderer = vtkSmartPointer<vtkRenderer>::New();
  m_Renderer->SetBackground(1.0, 1.0, 1.0);

  // Add a render window.
  m_RenderWindow = vtkSmartPointer<vtkRenderWindow>::New();
  m_RenderWindow->AddRenderer(m_Renderer);
  m_RenderWindow->SetSize(1000, 1000);
  m_RenderWindow->Render();
  m_RenderWindow->SetWindowName("Simulation Results Surface Slicer");

  // Add window interactor to use trackball and intercept key presses.
  //
  m_RenderWindowInteractor = vtkSmartPointer<vtkRenderWindowInteractor>::New();
  m_RenderWindowInteractor->SetRenderWindow(m_RenderWindow);

  m_InteractionStyle = vtkSmartPointer<MouseCenterlineInteractorStyle>::New();
  //m_InteractionStyle = vtkSmartPointer<MouseMeshInteractorStyle>::New();

  m_InteractionStyle->SetGraphics(this);
  m_RenderWindowInteractor->SetInteractorStyle(m_InteractionStyle);
  m_InteractionStyle->SetDefaultRenderer(m_Renderer);
  m_RenderWindowInteractor->SetInteractorStyle(m_InteractionStyle);

}

Mesh* Graphics::GetMesh()
{
  return m_Mesh;
}
void Graphics::SetMesh(Mesh* mesh)
{
  m_Mesh = mesh;
}

void Graphics::SetCenterlines(Centerlines& centerlines)
{
  m_Centerlines = centerlines;
  m_InteractionStyle->SetCenterlines(centerlines);
}

void Graphics::Start()
{
  //m_RenderWindowInteractor->Initialize();
  m_RenderWindowInteractor->Start();
}

//----------------
// CreateGeometry 
//----------------
//
vtkSmartPointer<vtkActor> Graphics::CreateGeometry(vtkSmartPointer<vtkPolyData> polyData)
{
  vtkSmartPointer<vtkDataSetMapper> mapper = vtkSmartPointer<vtkDataSetMapper>::New();
  mapper->SetInputData(polyData);
  mapper->ScalarVisibilityOff();
  vtkSmartPointer<vtkActor> actor = vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);
  return actor;
}

//-------------
// AddGeometry
//-------------
//
void Graphics::AddGeometry(vtkSmartPointer<vtkActor> geom)
{
  m_Renderer->AddActor(geom);
}

std::string Graphics::GetDataName() 
{
  return m_DataName;
}

void Graphics::SetDataName(std::string name) 
{
  if (m_Mesh != nullptr) {
    if (m_Mesh->HasData(name)) {
      m_DataName = name;
    } else {
      std::cout << "WARNING: The mesh does not have data named '" << name << "'" << std::endl;
    }
  } else {
      std::cout << "WARNING: The mesh is not set for graphics." << "'" << std::endl;
  }
}

// Create a circle
vtkSmartPointer<vtkActor> Graphics::CreateCircle()
{
  vtkSmartPointer<vtkRegularPolygonSource> polygonSource = vtkSmartPointer<vtkRegularPolygonSource>::New();
  polygonSource->SetNumberOfSides(50);
  polygonSource->SetRadius(5);
  polygonSource->SetCenter(0, 0, 0);

  vtkSmartPointer<vtkPolyDataMapper> mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
  mapper->SetInputConnection(polygonSource->GetOutputPort());;
      
  vtkSmartPointer<vtkActor> actor = vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);
  return actor;
}   

void Graphics::Refresh()
{
  m_RenderWindow->Render();
}

/////////////////////////////////////////////////////////////////////////
//            M o u s e M e s h I n t e r a c t o r S t y l e          //
/////////////////////////////////////////////////////////////////////////

// This class defines a mouse interactor used to select a surface
// element (triangle) and display its nodal data.

MouseMeshInteractorStyle::MouseMeshInteractorStyle()
{
  m_SelectedMapper = vtkSmartPointer<vtkDataSetMapper>::New();
  m_SelectedActor = vtkSmartPointer<vtkActor>::New();
  m_Colors = vtkSmartPointer<vtkNamedColors>::New();
}

void MouseMeshInteractorStyle::SetGraphics(Graphics* graphics)
{
  m_Graphics = graphics;
}

// Process a keyboard press event.
//
void MouseMeshInteractorStyle::OnKeyPress() 
{
  // Get the keypress.
  vtkRenderWindowInteractor *rwi = this->Interactor;
  std::string key = rwi->GetKeySym();
  std::cout << "Pressed key: " << key << std::endl;

  if ((key == "Escape") || (key == "q")) {
    exit(0);
  }
  else if (key == "s") {
    SelectCell();
  }

  // Forward events
  //vtkInteractorStyleTrackballCamera::OnLeftButtonDown();

}

//------------
// SelectCell
//------------
//
void MouseMeshInteractorStyle::SelectCell() 
{
  std::cout << "---------- Select Mesh Cell ----------" << std::endl;
  vtkSmartPointer<vtkNamedColors> colors = vtkSmartPointer<vtkNamedColors>::New();

  // Get the location of the click (in window coordinates)
  int* pos = this->GetInteractor()->GetEventPosition();

  vtkSmartPointer<vtkCellPicker> picker = vtkSmartPointer<vtkCellPicker>::New();
  picker->SetTolerance(0.0005);

  // Pick from this location.
  picker->Pick(pos[0], pos[1], 0, this->GetDefaultRenderer());

  double* worldPosition = picker->GetPickPosition();
  auto cellID = picker->GetCellId();

  if (cellID == -1) {
    return;
  }

  std::cout << "Pick position is: " << worldPosition[0] << " " << worldPosition[1] << " " << worldPosition[2] << endl;
  SelectMesh(cellID);
}

//-------------
// SelectMesh 
//-------------
//
void MouseMeshInteractorStyle::SelectMesh(int cellID)
{ 
  vtkSmartPointer<vtkIdTypeArray> ids = vtkSmartPointer<vtkIdTypeArray>::New();
  ids->SetNumberOfComponents(1);
  ids->InsertNextValue(cellID);

  vtkSmartPointer<vtkSelectionNode> selectionNode = vtkSmartPointer<vtkSelectionNode>::New();
  selectionNode->SetFieldType(vtkSelectionNode::CELL);
  selectionNode->SetContentType(vtkSelectionNode::INDICES);
  selectionNode->SetSelectionList(ids);

  vtkSmartPointer<vtkSelection> selection = vtkSmartPointer<vtkSelection>::New();
  selection->AddNode(selectionNode);

  if (m_Graphics->GetMesh()->IsSurface()) {
    SelectSurfaceMesh(cellID, selection);
  }

}

//-------------------
// SelectSurfaceMesh
//-------------------
//
void MouseMeshInteractorStyle::SelectSurfaceMesh(int cellID, vtkSmartPointer<vtkSelection> selection)
{ 
  auto mesh = m_Graphics->GetMesh();
  vtkPolyData* polydata = vtkPolyData::SafeDownCast(mesh->GetMesh());
  auto dataName = m_Graphics->GetDataName();
  auto data = mesh->GetDataArray(dataName);

  // Extract a list of cells from polydata.
  //
  vtkSmartPointer<vtkExtractSelection> extractSelection = vtkSmartPointer<vtkExtractSelection>::New();
  //vtkSmartPointer<vtkExtractSelectedPolyDataIds> extractSelection = vtkSmartPointer<vtkExtractSelectedPolyDataIds>::New();
  extractSelection->SetInputData(0, polydata); 
  extractSelection->SetInputData(1, selection);
  extractSelection->Update();

  // Get selected cells as a vtkPolyData.
  //
  vtkSmartPointer<vtkUnstructuredGrid> selected = vtkSmartPointer<vtkUnstructuredGrid>::New();
  //vtkSmartPointer<vtkPolyData> selected = vtkSmartPointer<vtkPolyData>::New();
  selected->ShallowCopy(extractSelection->GetOutput());
  std::cout << "There are " << selected->GetNumberOfPoints() << " points in the selection." << std::endl;
  std::cout << "There are " << selected->GetNumberOfCells() << " cells in the selection." << std::endl;

   // Print cell ID and connectivity.
   //
   vtkGenericCell* cell = vtkGenericCell::New();
   polydata->GetCell(cellID, cell);
   auto numPts = cell->GetNumberOfPoints();
   auto nodeIDs = vtkIntArray::SafeDownCast(polydata->GetPointData()->GetArray("GlobalNodeID"));
   auto elemIDs = vtkIntArray::SafeDownCast(polydata->GetCellData()->GetArray("GlobalElementID"));
   auto elemID = elemIDs->GetValue(cellID);
   std::cout << "Cell ID is: " << cellID << std::endl;
   std::cout << "  Elem ID: " << elemID << std::endl;
   std::cout << "  Number of cell points: " << numPts << std::endl;
   std::cout << "  Connectivity: ";
   for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
     auto id = cell->PointIds->GetId(pointInd);
     auto nodeID = nodeIDs->GetValue(id);
     std::cout << nodeID << " ";
   }
   std::cout << std::endl;

   // Print cell data.
   if (dataName != "") {
     std::cout << std::endl;
     std::cout << "Cell Data: " << std::endl;
     std::cout << "  Name: " << dataName << std::endl;
     auto numComp = data->GetNumberOfComponents();

     for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
       auto id = cell->PointIds->GetId(pointInd);
       auto nodeID = nodeIDs->GetValue(id);
       if (numComp == 1) { 
         double value = data->GetValue(id);
         std::cout << "  ID: " << nodeID << "   Value: " << value << std::endl;
       } else if (numComp == 3) { 
         auto v1 = data->GetComponent(id, 0);
         auto v2 = data->GetComponent(id, 1);
         auto v3 = data->GetComponent(id, 2);
         std::cout << "  Node ID: " << nodeID << "   Values: " << v1 << " " << v2 << " " << v3 << std::endl;
       }
     }
   }

  // Display the selected cell.
  //
  m_SelectedMapper->SetInputData(selected);
  m_SelectedMapper->ScalarVisibilityOff();
  m_SelectedActor->SetMapper(m_SelectedMapper);
  //selectedActor->GetProperty()->EdgeVisibilityOn();
  //m_SelectedActor->GetProperty()->SetRepresentationToWireframe();
  m_SelectedActor->GetProperty()->SetColor(m_Colors->GetColor3d("Green").GetData());
  m_SelectedActor->GetProperty()->SetLineWidth(4);

  this->Interactor->GetRenderWindow()->GetRenderers()->GetFirstRenderer()->AddActor(m_SelectedActor);
  this->Interactor->GetRenderWindow()->Render();
}

//void MouseInteractorStyle::OnLeftButtonDown()
//{
//}

vtkStandardNewMacro(MouseMeshInteractorStyle);

/////////////////////////////////////////////////////////////////////////////////////
//            M o u s e C e n t e r l i n e I n t e r a c t o r S t y l e          //
/////////////////////////////////////////////////////////////////////////////////////

// This class defines a mouse interactor used to create a slice at a selected
// point on a surface centerline.

MouseCenterlineInteractorStyle::MouseCenterlineInteractorStyle()
{
  startSelected = false;
}

void MouseCenterlineInteractorStyle::SetCenterlines(Centerlines& centerlines)
{
  m_Centerlines = centerlines;
}

void MouseCenterlineInteractorStyle::SetGraphics(Graphics* graphics)
{
  m_Graphics = graphics;
}

//------------
// OnKeyPress
//------------
//
void MouseCenterlineInteractorStyle::OnKeyPress() 
{
  // Get the keypress.
  vtkRenderWindowInteractor *rwi = this->Interactor;
  std::string key = rwi->GetKeySym();

  // Output the key that was pressed.
  //std::cout << "Pressed " << key << std::endl;

  // Undo. Removes the last slice.
  if (key == "u") {
    auto mesh = m_Graphics->GetMesh();
    mesh->UndoSlice();

  // Write slices to a file.
  } else if (key == "w") {
    auto mesh = m_Graphics->GetMesh();
    mesh->WriteSlices();

  // Quit.
  } else if ((key == "Escape") || (key == "q")) {
    exit(0);
  }

  vtkInteractorStyleTrackballCamera::OnKeyPress();
}

//------------------
// SelectCenterline
//------------------
//
void MouseCenterlineInteractorStyle::OnLeftButtonDown() 
//void MouseCenterlineInteractorStyle::SelectCenterline() 
{
  // Pick current screen location.
  //
  int* clickPos = this->GetInteractor()->GetEventPosition();
  vtkSmartPointer<vtkPropPicker> picker = vtkSmartPointer<vtkPropPicker>::New();
  picker->Pick(clickPos[0], clickPos[1], 0, this->GetDefaultRenderer());

  if (picker->GetActor() != nullptr) {
    double* pos = picker->GetPickPosition();
    //std::cout << "Pick position (world coordinates) is: " << pos[0] << " " << pos[1] << " " << pos[2] << std::endl;
    //std::cout << "Picked actor: " << picker->GetActor() << std::endl;

    // Create a sphere.
    if (startSphere == nullptr) {
      startSphere = vtkSmartPointer<vtkSphereSource>::New();
      vtkSmartPointer<vtkPolyDataMapper> mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
      mapper->SetInputConnection(startSphere->GetOutputPort());
      vtkSmartPointer<vtkActor> actor = vtkSmartPointer<vtkActor>::New();
      actor->SetMapper(mapper);
      actor->GetProperty()->SetColor(1.0, 0.0, 0.0);
      //this->GetDefaultRenderer()->AddActor(actor);
    }

    if (startPlane == nullptr) {
      startPlane = vtkSmartPointer<vtkPlaneSource>::New();
      vtkSmartPointer<vtkPolyDataMapper> pmapper = vtkSmartPointer<vtkPolyDataMapper>::New();
      pmapper->SetInputConnection(startPlane->GetOutputPort());
      vtkSmartPointer<vtkActor> startPlaneActor = vtkSmartPointer<vtkActor>::New();
      startPlaneActor->SetMapper(pmapper);
      startPlaneActor->GetProperty()->SetColor(1.0, 0.0, 0.0);
      //this->GetDefaultRenderer()->AddActor(startPlaneActor);
    }

    double radius = 0.1;
    double inscribedRadius;
    double normal[3], tangent[3], binormal[3], p1[3];
    double planeWidth, origin[3], point1[3], point2[3], vec1[3], vec2[3];
    int index;
    int cellID;

    // Get centerline data at the picked point.
    m_Centerlines.locate_cell(pos, index, cellID, inscribedRadius, normal, tangent);

    startSphere->SetCenter(pos[0], pos[1], pos[2]);
    startSphere->SetRadius(radius);
    //startSphere->SetRadius(radius);
    //
    planeWidth = 4.0*inscribedRadius;
    //planeWidth = 2.0*inscribedRadius;
    vtkMath::Cross(normal, tangent, binormal);
    vtkMath::Normalize(binormal);
    vtkMath::Normalize(tangent);
    //vtkMath::Perpendiculars(normal, vec1, vec2, vtkMath::Pi()/2.0);
    vtkMath::Perpendiculars(normal, vec1, vec2, 0.0);

    for (int i = 0; i < 3; i++) {
      origin[i] = pos[i] - planeWidth/2.0*tangent[i] - planeWidth/2.0*normal[i];
      point1[i] = origin[i] + planeWidth*tangent[i];
      point2[i] = origin[i] + planeWidth*normal[i];
      p1[i] = pos[i] + 0.5*normal[i];
      p1[i] = pos[i] + 0.5*tangent[i];
    }
    startPlane->SetCenter(pos[0], pos[1], pos[2]);
    startPlane->SetNormal(tangent[0], tangent[1], tangent[2]);

    auto mesh = m_Graphics->GetMesh();
    auto dataName = m_Graphics->GetDataName();

    // Extract a slice from the mesh.
    mesh->SlicePlane(index, cellID, dataName, pos, tangent);
  }

  this->Interactor->GetRenderWindow()->Render();

  // Forward events
  vtkInteractorStyleTrackballCamera::OnLeftButtonDown();
}

vtkStandardNewMacro(MouseCenterlineInteractorStyle);

File: image-slicer/python/graphics.py
#!/usr/bin/env python

import logging
from manage import get_logger_name
import numpy as np
from os import path
from path import Path
import vtk

class MouseInteractorStyle(vtk.vtkInteractorStyleTrackballCamera):

    def __init__(self, graphics):
        #self.AddObserver("LeftButtonPressEvent", self.leftButtonPressEvent)
        self.AddObserver("KeyPressEvent", self.onKeyPressEvent)
        self.AddObserver("CharEvent", self.onCharEvent)
        self.graphics = graphics
        self.selected_points = []
        self.sphere = None

    def leftButtonPressEvent(self, obj, event):
        ''' 
        Process left mouse button press.
        '''
        print("leftButtonPressEvent: ")
        clickPos = self.GetInteractor().GetEventPosition()

        # vtk.vtkCellPicker() does not select path lines.
        #picker = vtk.vtkCellPicker()
        picker = vtk.vtkPropPicker()
        picker.Pick(clickPos[0], clickPos[1], 0, self.renderer)
        #picker.SetTolerance(0.0001)
        if picker.GetActor() == None:
            return

        position = picker.GetPickPosition()
        print("  position: " + str(position))

        if self.sphere == None:
            sphere = vtk.vtkSphereSource()
            sphere.SetCenter(position[0], position[1], position[2])
            sphere.SetRadius(0.1)
            sphere.Update()
            polydata = sphere.GetOutput()
            mapper = vtk.vtkPolyDataMapper()
            mapper.SetInputData(polydata)
            mapper.ScalarVisibilityOff()
            actor = vtk.vtkActor()
            actor.SetMapper(mapper)
            actor.GetProperty().SetRepresentationToWireframe()
            actor.GetProperty().SetColor(0.0, 1.0, 0.0)
            self.graphics.add_actor(actor)
            self.sphere = sphere

        self.sphere.SetCenter(position[0], position[1], position[2])
        self.sphere.Update()

        ## Get path data at the selected point.
        for path in self.graphics.paths:
            path_data = path.select(position)
            print("  path data: ")
            print("    id: %s" %  path_data.id)
            print("    index: %d" %  path_data.index)
            print("    point: %s" %  str(path_data.point))
            print("    tangent: %s" %  str(path_data.tangent))
            print("    rotation: %s" %  str(path_data.rotation))

            # Show tangent.
            s = 1.0
            pt1 = path_data.point
            tangent = path_data.tangent
            tangent = np.array(path_data.tangent)
            pt2 = [(pt1[j]+s*tangent[j]) for j in range(0,3)] 
            self.graphics.add_line(pt1, pt2, width=5)

            # Show normal.
            normal = np.array(path_data.rotation)
            pt2 = [(pt1[j]+s*normal[j]) for j in range(0,3)] 
            self.graphics.add_line(pt1, pt2, color=[0.0,1.0,0.0], width=5)

            # Show binormal.
            binormal = np.cross(tangent, normal)
            pt2 = [(pt1[j]+s*binormal[j]) for j in range(0,3)] 
            self.graphics.add_line(pt1, pt2, color=[1.0,1.0,0.0], width=5)

            self.graphics.image.extract_slice(pt1, tangent, normal, binormal)

        self.graphics.window.Render()

        return

    def onCharEvent(self, renderer, event):
        """
        Process an on char event.

        This is used to prevent passing the shortcut key 'w' to vtk which we use
        to write selected results and vtk uses to switch to wireframe display. 
        """
        key = self.GetInteractor().GetKeySym()
        if (key != 'w'):
            self.OnChar()
  
    def onKeyPressEvent(self, renderer, event):
        """
        Process a key press event.
        """
        key = self.GetInteractor().GetKeySym()

        if (key == 's'):
            self.leftButtonPressEvent(None, event)
        elif (key == 'f'):
            self.fix()

    #__def onKeyPressEvent

#__class MouseInteractorStyle

class Graphics(object):

    def __init__(self):
        self.renderer = None
        self.window = None
        self.interactor = None
        self.image = None
        self.logger = logging.getLogger(get_logger_name())
        self.initialize_graphics()

    def add_actor(self, actor):
        self.renderer.AddActor(actor)

    def initialize_graphics(self):
        """ Create renderer and graphics window.
        """
        self.renderer = vtk.vtkRenderer()
        self.window = vtk.vtkRenderWindow()
        self.window.AddRenderer(self.renderer)
        self.renderer.SetBackground(0.5, 0.5, 0.5)
        self.window.SetSize(1000, 1000)
        #self.window.Render()

        # Create a trackball interacter to transoform the geometry using the mouse.
        self.interactor = vtk.vtkRenderWindowInteractor()
        self.interactor.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
        self.interactor.SetRenderWindow(self.window)

        # Add the custom style.
        style = MouseInteractorStyle(self)
        style.renderer = self.renderer
        style.graphics = self
        self.interactor.SetInteractorStyle(style)
        style.SetCurrentRenderer(self.renderer)

    def add_line(self, pt1, pt2, color=[1.0, 1.0, 1.0], width=2):
        lineSource = vtk.vtkLineSource()
        lineSource.SetPoint1(pt1);
        lineSource.SetPoint2(pt2)
        lineSource.Update()
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputData(lineSource.GetOutput())
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        actor.GetProperty().SetLineWidth(width)
        actor.GetProperty().SetColor(color[0], color[1], color[2])
        self.add_actor(actor)

    def add_sphere(self, center, radius=1.0, color=[1.0,1.0,1.0]):
        sphere = vtk.vtkSphereSource()
        sphere.SetCenter(center[0], center[1], center[2])
        sphere.SetRadius(radius)
        sphere.Update()
        polydata = sphere.GetOutput()
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputData(polydata)
        mapper.ScalarVisibilityOff()
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        #actor.GetProperty().SetRepresentationToWireframe()
        actor.GetProperty().SetColor(color[0], color[1], color[2])
        self.renderer.AddActor(actor)

    def add_graphics_geometry(self, poly_data, color, sphere=False):
        gr_geom = self.create_graphics_geometry(poly_data, sphere)
        gr_geom.GetProperty().SetColor(color[0], color[1], color[2])
        #gr_geom.GetProperty().SetRepresentationToWireframe()
        gr_geom.GetProperty().SetEdgeColor(0.0, 0.0, 0.0)
        gr_geom.GetProperty().EdgeVisibilityOn()
        self.renderer.AddActor(gr_geom)
        self.window.Render()
        self.window.SetWindowName("Extract Faces")
        return gr_geom 

    def show(self):
        self.window.Render()
        self.window.SetWindowName("Image Slice")
        self.interactor.Start()

class ClickInteractorStyle(vtk.vtkInteractorStyleTrackballCamera):
    def __init__(self, graphics):
        self.graphics = graphics
        self.AddObserver("KeyPressEvent", self.onKeyPressEvent)

    def onKeyPressEvent(self, renderer, event):        
        key = self.GetInteractor().GetKeySym()
        if key == 'c':
            pass 


